base/main_test_bucket_pool.go

```diff
@@ -0,0 +1,650 @@
+		ctx := bucketNameCtx(context.Background(), testBucketName)
+
+		var bucketExists bool
+		for _, b := range existingBuckets {
+			if testBucketName == b.Name {
+				tbp.Logf(ctx, "Skipping InsertBucket... Bucket already exists")
+				bucketExists = true
+			}
+		}
+
+		// Bucket creation takes a few seconds for each bucket,
+		// so create and wait for readiness concurrently.
+		go func(i int, bucketExists bool) {
+			if !bucketExists {
+				tbp.Logf(ctx, "Creating new test bucket")
+				err := tbp.clusterMgr.InsertBucket(&gocb.BucketSettings{
+					Name:          testBucketName,
+					Quota:         bucketQuotaMB,
+					Type:          gocb.Couchbase,
+					FlushEnabled:  true,
+					IndexReplicas: false,
+					Replicas:      0,
+				})
+				if err != nil {
+					tbp.Logf(ctx, "Couldn't create test bucket: %v", err)
+					os.Exit(1)
+				}
+			}
+
+			b, err := tbp.openTestBucket(tbpBucketName(testBucketName), CreateSleeperFunc(5*numBuckets, 1000))
+			if err != nil {
+				tbp.Logf(ctx, "Timed out trying to open new bucket: %v", err)
+				os.Exit(1)
+			}
+			openBuckets[i] = b
+
+			wg.Done()
+		}(i, bucketExists)
+	}
+
+	// wait for the async bucket creation and opening of buckets to finish
+	wg.Wait()
+
+	// All the buckets are created and opened, so now we can perform some synchronous setup (e.g. Creating GSI indexes)
+	for i := 0; i < numBuckets; i++ {
+		testBucketName := tbpBucketNamePrefix + strconv.Itoa(i)
+		ctx := bucketNameCtx(context.Background(), testBucketName)
+
+		tbp.Logf(ctx, "running bucketInitFunc")
+		b := openBuckets[i]
+
+		if err, _ := RetryLoop(b.GetName()+"bucketInitRetry", func() (bool, error, interface{}) {
+			tbp.Logf(ctx, "Running bucket through init function")
+			err = bucketInitFunc(ctx, b, tbp)
+			if err != nil {
+				tbp.Logf(ctx, "Couldn't init bucket, got error: %v - Retrying", err)
+				return true, err, nil
+			}
+			return false, nil, nil
+		}, CreateSleeperFunc(5, 1000)); err != nil {
+			tbp.Logf(ctx, "Couldn't init bucket, got error: %v - Aborting", err)
+			os.Exit(1)
+		}
+
+		b.Close()
+		tbp.addBucketToReadierQueue(ctx, tbpBucketName(testBucketName))
+	}
+
+	return nil
+}
+
+// bucketReadierWorker reads a channel of "dirty" buckets (bucketReadierQueue), does something to get them ready, and then puts them back into the pool.
+// The mechanism for getting the bucket ready can vary by package being tested (for instance, a package not requiring views or GSI can use FlushBucketEmptierFunc)
+// A package requiring views or GSI, will need to pass in the db.ViewsAndGSIBucketReadier function.
+func (tbp *GocbTestBucketPool) bucketReadierWorker(ctx context.Context, bucketReadierFunc TBPBucketReadierFunc) {
+	tbp.Logf(context.Background(), "Starting bucketReadier")
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			tbp.Logf(context.Background(), "bucketReadier got ctx cancelled")
+			break loop
+
+		case testBucketName := <-tbp.bucketReadierQueue:
```

Comment:
The readierQueue is taking bucket names, and then reopening the recently closed bucket so that it can invoke the bucketReadierFunc with an open bucket.

Is there an option to make bucketReadierQueue a channel of open buckets, and have GetTestBucketAndSpec pass the open bucket instead of closing and passing the name?

Reply:
It was doing this at one point, but I decided that it can't be guaranteed that tests are going to be keeping buckets open in all cases (e.g. indirect bucket closure via `DatabaseContext.Close()`)